### ISSUE TITLE
[RFC] Refactoring Bolt Controllers, and Middlewares

### ISSUE THREAD

**Issue by CarsonF** [bolt/bolt#3537]
## TL;DR
This is a total refactor of controller and routing logic. 

This is a joint working-group lead by myself with input, commits and guidance from @GawainLynch to tackle refactoring of Bolt’s routing and controller functionality to ensure consistency and maintainability.

## Tracking Branch
The branch referenced in this RFC can be [examined here](https://github.com/bolt/bolt/compare/master...CarsonF:feature/controller-refactor)

**NOTE 1:** Due to a large amount of code being moved out of one class and (generally speaking) to new classes, the `diff` is **massive**, and hard to read. For review we recommend pulling the branch, and either reviewing individual commits, or reviewing actual files in-place.

**NOTE 2:** The size of this as a changeset is largely dictated by the amount of code required to be changed in order to maintain full test coverage for merging.

## Reserving Merge Priority
We are requesting — should the approach in this branch be agreed upon — a first-in-line merge into `master`, immediately after `release/2.2` is branched. 

This branch shouldn’t conflict with @rossriley’s `feature/entity-manager-refactor` branch (past unforeseen one-liners).

However, there are incoming changes by others that will be affected by this and we would prefer that authors get credit for their additions rather than us porting those changes on their behalf. 

An early merge will also allow for the greatest amount of test coverage, and feedback for these changes.

## Tests
As of this RFC, all existing unit and acceptance tests pass. 

We have put a significant amount of effort into improving controller tests code, and we have increased coverage where possible (mainly with respect to asynchronous routes).

## Problems
1. Bolt’s backend controllers are monolithic and bloated
1. The Frontend controller largely static and inflexible
1. The routing configuration unextendable
1. Controllers were cluttered with DI service calls that can’t be type hinted
1. There was a lack of consistency in the way service calls we used/implemented
1. Before/after methods were confusing, and unorganized
1. Controller mounting order was hard coded, which made it impossible for extensions to define certain routes. e.g. `/foo`
1. Responses returned from Frontend routes couldn’t be modified.
1. Data fetched in Frontend routes couldn't be used by after middlewares
1. Application class is cluttered with request/response logic. Application should register Service Providers, nothing more.
1. Response object outcomes are unpredictable and hard-wired into setup logic, instead of being clearly defined in the routing logic
1. The `routing.yml` file referenced class names directly, so these controllers couldn’t be changed/extended.
1. Kittens were getting upset… Won’t someone please think of the kittens!!

## Proposed Approach

1. Maintain near-full backwards compatibility
1. Controllers should extend a `Base` class that has defined “shortcut” methods to clean up DI service calls. [See Symfony’s base controller class](https://github.com/symfony/symfony/blob/2.7/src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php) 
1. Controllers broken out into sub-namespaced sub-classes, by their functional subject
1. Before/after methods and handlers in Application should be moved to listener classes by their functional subject. e.g. `MaintenanceModeListener`
1. Controllers should be mounted on an event, so additional controllers can be mounted at any priority
1. Controllers should be services in DI, so they can be changed/extended and called in other places if needed
1. Dynamic/YAML base routing (`routing.yml`) should reference controller service names instead of named classes. e.g. `controller.frontend:record`
1. Dynamic/YAML base routing handled, with full backwards compatibility, via a `ConfigurableBase` class, that extends `Base`, which parses the YAML array to routes and leverages the Symfony Callback Resolver to identify the associated controller and middlewares

AKA: Moar better.

## Controller Mounting
Controllers are now collectively mounted/initialized in `Bolt\Provider\ControllerServiceProvider` and exist as DI objects, e.g. `$app['controller.foo']`

The general code path is:
* During the `Application::initialize()` a `ControllerServiceProvider` is registered
* In the `ControllerServiceProvider::register()` function the controller is added to the DI container via a closure, e.g.
```php
$app['controller.foo] = $app->share(function () {
   return new Controller\Foo();
});
```
* During `ControllerServiceProvider::boot()`:
  * `ControllerServiceProvider` adds itself as an event subscriber
  * `ControllerServiceProvider` dispatches a mount event to notify which controllers should be mounted
  * `ControllerServiceProvider`, being an EventSubscriber, gets this event dispatched to `ControllerServiceProvider::mount()` (this method replaces `Application::initMountpoints()`)
* `ControllerServiceProvider::mount()` then calls `Bolt\Events\MountEvent::mount($pathPrefix, $app['controller.foo'], $priority)` which adds the controller object from the DI container to the event:
```php
$event->mount('/my-route', $app['controller.foo'], 0);
```
* After the event is done dispatching, `MountEvent::finish` is called which:
  * Sorts the controllers based on priorities given
  * Actually mounts them to the `Bolt\Application`

## Middlewares (Request and Response Handling)
The changes to controller handling up until this point are the foundation to allow us to build on the middleware handling. Some of this is still work in progress, but the foundations are critical and hence what we are mainly proposing in this RFC. Some aspects are subject to change on peer-review.

The handlers added in both `Application::before()` and `Application::after()` have been, or are in the process of being moved to listener classes such as:
  * `MaintenanceModeListener`

## Response Handling

As Response objects are now manipulated using kernel events
* Do not use functions like these:
  * session_*()
  * headers_*()
  * setcookie()

This will create consistent and manageable response behaviour.

Where cookie or header manipulation is required, it should be done via the `Response` object in an _after middleware_. After middlewares can be created a couple different ways:
  * Calling `$app->after()` with an anonymous function.
  * Creating a “Listener” class that implements `EventSubscriberInterface`. This class should be attached to the EventDispatcher at boot.

Core will be primarily using the second method as this class can be put into DI and extended/modified by others.

### Backend
The backend controllers are now:
* Bolt\Controller\Backend\Authentication
  * A combination of old backend before() logic and Bolt\Users
* Bolt\Controller\Backend\General
  * Generic routes for the backend such as the dashboard
* Bolt\Controller\Backend\Database
  * Database maintenance routes
* Bolt\Controller\Backend\Extend
  * The old Bolt\Controllers\Extend class 
* Bolt\Controller\Backend\FileManager
  * File maintenance and management routes
* Bolt\Controller\Backend\Log
  * System & content log maintenance and management routes
* Bolt\Controller\Backend\Records
  * A massive cleanup and refactor of the Contenttype record management routes
  * Backend::editcontent() was nearly 400 lines of code, and is now 20 calling out to private helper functions
* Bolt\Controller\Backend\Upload
  * Moved from Bolt\Controllers\Upload
* Bolt\Controller\Backend\Users
  * User maintenance and management routes

#### Before

| Controller                              | SLoC | Total |
| ----------------------------------------|:----:|------:|
| Bolt\Controllers\Backend                | 1409 |  2191 |
| Bolt\Controllers\Extend                 |  228 |   420 |
| Bolt\Controllers\Upload                 |  175 |   230 |

#### After (as at 2015-05-07)

| Controller                              | SLoC | Total |
| ----------------------------------------|-----:|------:|
| Bolt\Controller\Backend\Authentication  |   79 |   121 |
| Bolt\Controller\Backend\Database        |   50 |    99 |
| Bolt\Controller\Backend\Extend          |  181 |   366 |
| Bolt\Controller\Backend\FileManager     |  256 |   373 |
| Bolt\Controller\Backend\General         |  192 |   339 |
| Bolt\Controller\Backend\Log             |  132 |   236 |
| Bolt\Controller\Backend\Records         |  203 |   369 |
| Bolt\Controller\Backend\Upload          |  125 |   240 |
| Bolt\Controller\Backend\Users           |  402 |   628 |

### Async
As with the old `Backend` controller, the `Async` controller has been broken into namespaced subclasses:
* Bolt\Controller\Async\FilesystemManager
* Bolt\Controller\Async\General
* Bolt\Controller\Async\Stack
* Bolt\Controller\Async\SystemTests

## Access Control
Functionality for access control has been split out of Bolt\Users and into `Bolt\AccessControl\Authentication` additionally the permissions classes are now in this namespace.

Primarily this puts access control functions into the HTTP request space, and out of application set-up where is caused scope access issues previously.

#### Controller Route Function Naming
Route functions in controllers are currently prefixed with the word `action`. 

This has been done as a development indicator to highlight *“It’s a route”*, make things easy to `grep` or `ack` while in flux, and it is not (currently) expected to remain for final release. Put more simply, blame the :koala:, he insists on using the CLI and vim.

Symfony currently document route functions with `Action` as a suffix, and we were unable to come up with a workable *final* decision, but for the early development cycle it made sense to us. Feedback and suggestions are **very much** desired.

## Extensions

Extensions should create routes inside a controller class extending Base, not directly in the `Bolt\Extensions\AuthorName\Extension` class, e.g.:

```php
// Extension.php
public function initialize()
{
    $this->boot($this->app);
}

public function boot(Application $app)
{
    /** @var \Symfony\Component\EventDispatcher\EventDispatcherInterface $dispatcher */
    $dispatcher = $app['dispatcher'];
    $dispatcher->addSubscriber($this);
}

public function mount(MountEvent $event)
{
    $app = $event->getApp();
    $this->app->mount('/', new Controller());
}

public static function getSubscribedEvents()
{
    return [
       ControllerEvents::MOUNT => 'mount',
    ];
}

```
**NOTE:** This is an example to illustrate the use of an implementation that is yet to be finalised, subject to change before 2.3 final, and should be considered as **for illustration purposes only**.

### Backend — `$app['htmlsnippets']`
The application boolean parameter `$app['htmlsnippets']` has been removed and replaced with a `Request` parameter.

Extensions requiring backend snippets will simply need to set the `Request` parameter ‘allow_snippets’ to true. If the current request uses the extension’s controller method the attribute can be set there. If not, it can be set in a `before` middleware. e.g.
````php
// Controller.php
public function addRoutes(ControllerCollection $controllers)
{
    $controllers->before(function (Request $request) {
        $request->attributes->set('allow_snippets', true);
    });
}
``` 

**WARNING NOTE:** If one extension sets this request attribute in the `Application` before middleware, all extensions installed will allow snippets in backend. This should be avoided and extensions should instead implement their own before middleware. The recommended way is overriding the `before()` method of `Bolt\Controller\Base`.


#### Justification
We are proposing this as a breaking change for the following reasons:
  * Bolt v2.2 is going to be a Long Term Support release, and this is targeting v2.3
  * The benefit of making it outweighs the workarounds, based on the points laid out in our reasoning above
  * We believe that the number of adversely affected users/extensions is tiny
    * BoltBB
    * Members
    * MenuEditor
  * A big warning in release notes can mitigate problems and provide choice

## Follow-up work (post-merge)
* Extension initialisation
* Further clean up of controller and form logic
* :leaves: for the :koala:

## Conclusion
Can I has now? 

Soon, but you can grab a coffee and re-read this whole RFC… you know you want to! 

#### Feedback and constructive criticism desired!

**Comment by bobdenotter** [comment]
Awesome! 

One little nitpick: If this has breaking changes, shouldn't we bump '2.3' to '3.0' ?

**Comment by GawainLynch** [comment]
We feel that's not the best approach, it's also not as significant as dropping PHP 5.3 :grin: 

The break is a small one and corner case, and I can help @bacbos make the change in MenuEditor if he needs, and the other two are mind.

One advantage of staying in a 2.x is that we can spend the rest of the series marking out what we want/need to change and have a clear view when we start on 3.0

**Comment by GawainLynch** [comment]
For anyone wanting to pull this branch into a local, and are not a master of the git-fu ways:
```sh
git checkout -b feature/controller-refactor master
git pull https://github.com/CarsonF/bolt.git feature/controller-refactor
```

**Comment by GawainLynch** [comment]
@bobdenotter just an additional thought on the BC break. It only affects backend, which is something we've not guaranteed BC on anyway.

**Comment by rossriley** [comment]
Absolutely awesome work.

**Comment by xiaohutai** [comment]
Wow, that's insane :+1:

**Comment by rixbeck** [comment]
I like that direction
